File: frontend/src/api/workflow.ts
import { get, post, put, del } from '@/utils/request'
import { WORKFLOW_NODE_TYPES } from './workflowContract'
import type { WorkflowNodeType, WorkflowStatus } from './workflowContract'

/**
 * Workflow orchestration client.
 *
 * Node `type` doubles as the DSL `component_name`. The DSL keeps two
 * synchronized views:
 *  - `graph`: canvas layout (nodes with positions + edges), consumed by the
 *    vue-flow editor;
 *  - `components`: execution topology (upstream/downstream + params),
 *    consumed by the backend engine.
 *
 * Param keys inside `components[id].obj.params` MUST match the engine
 * registry verbatim (snake_case, e.g. `kb_ids`, `system_prompt`,
 * `top_k`) — the engine reads them by literal key and fails on missing
 * required ones. Typed interfaces below document that contract; the DSL
 * itself keeps `params` as a loose record because not every field is set.
 *
 * The pure type/constant contract (WorkflowNodeType, WORKFLOW_NODE_TYPES,
 * ...) lives in ./workflowContract (zero imports); this module re-exports
 * it so existing importers keep working.
 */
export type {
  WorkflowNodeType,
  WorkflowStatus,
} from './workflowContract'
export { WORKFLOW_NODE_TYPES, NODE_OUTPUT_PARAMS } from './workflowContract'

// ---- Typed param shapes (per-node property forms) --------------------

export interface TemplateOp {
  op: 'upper' | 'lower' | 'trim' | 'replace' | 'regex_extract'
  from?: string
  to?: string
  pattern?: string
  group?: number
}

/** One comparison inside a Switch case group. */
export interface SwitchCondition {
  /** Left operand template, e.g. "{llm@content}". */
  ref: string
  /** Operator id (eq/ne/contains/not_contains/starts_with/ends_with/empty/not_empty/gt/gte/lt/lte/regex/in/not_in). */
  op: string
  /** Right operand literal/template; ignored by empty/not_empty. */
  value: string
}

/** One Switch routing rule: condition group + target node. */
export interface SwitchCaseGroup {
  conditions: SwitchCondition[]
  logic: 'and' | 'or'
  to: string
}

/** Start-node input form field declaration. */
export interface StartField {
  name: string
  label?: string
  type: 'text' | 'paragraph' | 'number' | 'select'
  required?: boolean
  default?: string
  /** select choices */
  options?: string[]
}

export interface VariableRef {
  name: string
  ref: string
}

export interface WFPosition {
  x: number
  y: number
}

export interface WFNode {
  id: string
  type: WorkflowNodeType
  position: WFPosition
  data?: Record<string, unknown>
}

export interface WFEdge {
  id: string
  source: string
  target: string
  sourceHandle?: string
}

export interface WFComponent {
  obj: {
    component_name: string
    params: Record<string, unknown>
  }
  upstream: string[]
  downstream: string[]
}

export interface WorkflowDSL {
  version: 1
  graph: {
    nodes: WFNode[]
    edges: WFEdge[]
  }
  components: Record<string, WFComponent>
  variables?: Record<string, unknown>
}

export interface Workflow {
  id: string
  tenant_id?: number
  creator_id?: string
  name: string
  description?: string
  dsl: WorkflowDSL
  status: WorkflowStatus
  version?: number
  created_at?: string
  updated_at?: string
}

export interface WorkflowListResponse {
  success: boolean
  message?: string
  // Backend wraps the page: { workflows: Workflow[], total, page, page_size }.
  data: {
    workflows: Workflow[]
    total: number
    page?: number
    page_size?: number
  }
}

export interface WorkflowMutationResponse {
  success: boolean
  data?: Workflow
  message?: string
}

export const listWorkflows = (): Promise<WorkflowListResponse> => get('/api/v1/workflows')

export const getWorkflow = (id: string): Promise<{ success: boolean; data?: Workflow; message?: string }> =>
  get(`/api/v1/workflows/${id}`)

export const createWorkflow = (payload: {
  name: string
  description?: string
  dsl?: WorkflowDSL
}): Promise<WorkflowMutationResponse> => post('/api/v1/workflows', payload)

export const updateWorkflow = (
  id: string,
  payload: { name?: string; description?: string; dsl?: WorkflowDSL },
): Promise<WorkflowMutationResponse> => put(`/api/v1/workflows/${id}`, payload)

export const deleteWorkflow = (id: string): Promise<{ success: boolean }> => del(`/api/v1/workflows/${id}`)

// ---------------------------------------------------------------------------
// Run execution + progress (consumes the stage-2 backend contract).
//
// Envelope note: create-run answers with a bare { run } object (sync 200 /
// async 202 / failed-with-record 200), while run history follows the
// repository-wide { success, data } envelope — the shapes are kept apart in
// the response types below instead of papered over.
// ---------------------------------------------------------------------------

export type WorkflowRunStatus = 'pending' | 'running' | 'succeeded' | 'failed' | 'cancelled'

export interface WorkflowRunOutput {
  answer?: string
  path?: string[]
  outputs?: Record<string, Record<string, unknown>>
}

/** One persisted per-node record of a run's trace (run-detail endpoint). */
export interface WorkflowRunTraceEntry {
  node_id: string
  /** Component name ("LLM", "Retrieval", ...); empty when unknown. */
  kind?: string
  /** Terminal phase of this attempt: finished | failed. */
  phase: string
  duration_ms?: number
  outputs?: Record<string, unknown>
  error?: string
  /** true when restored from a checkpoint (resume), duration 0. */
  replayed?: boolean
}

export interface WorkflowRun {
  id: string
  tenant_id?: number
  workflow_id: string
  status: WorkflowRunStatus
  input?: unknown
  output?: WorkflowRunOutput | null
  /** Present on run-detail responses; history list rows omit it. */
  trace?: WorkflowRunTraceEntry[] | null
  error?: string
  created_at?: string
  updated_at?: string
}

/** One SSE frame of GET /workflows/:id/runs/:runId/events. */
export interface WorkflowRunEventFrame {
  workflow_id: string
  run_id: string
  kind: 'node' | 'run'
  /** Set for kind=node frames; matches the canvas node id. */
  node_id?: string
  /** node frames: started|finished|failed · run frames: terminal run status. */
  phase: string
  error?: string
  duration_ms?: number
  /** Node outputs on finished frames (run-debugging payload). */
  outputs?: Record<string, unknown>
  /** true when the finished frame replayed from a checkpoint (resume). */
  replayed?: boolean
  status?: WorkflowRunStatus
}

export interface WorkflowRunResponse {
  success?: boolean
  run?: WorkflowRun
  message?: string
}

export interface WorkflowRunListResponse {
  success: boolean
  data?: { runs: WorkflowRun[]; total: number }
  message?: string
}

export const runWorkflow = (
  id: string,
  payload: { query: string; files?: string[]; inputs?: Record<string, unknown>; async?: boolean },
): Promise<WorkflowRunResponse> => post(`/api/v1/workflows/${id}/runs`, payload)

/**
 * Cancel a pending/running run (best-effort engine stop + async-task
 * dequeue). Terminal runs answer idempotently with their current state —
 * cancelling an already-finished run is not an error. The SSE stream (if
 * attached) receives a kind=run phase=cancelled terminal frame, then closes.
 */
export const cancelWorkflowRun = (workflowId: string, runId: string): Promise<WorkflowRunResponse> =>
  post(`/api/v1/workflows/${workflowId}/runs/${runId}/cancel`)

/**
 * Resume a failed run from its checkpoint: the run row flips back to
 * running and re-executes asynchronously, skipping nodes whose outputs are
 * already checkpointed. Only status=failed runs are resumable — the server
 * answers 409 for other terminal states and 404 for unknown ids; a failed
 * run without checkpoint state simply re-executes from the start.
 */
export const resumeWorkflowRun = (workflowId: string, runId: string): Promise<WorkflowRunResponse> =>
  post(`/api/v1/workflows/${workflowId}/runs/${runId}/resume`)

export const listWorkflowRuns = (id: string): Promise<WorkflowRunListResponse> =>
  get(`/api/v1/workflows/${id}/runs`)

/**
 * Run detail — the full row including the per-node trace (execution order).
 * History list rows deliberately omit `trace`; this endpoint is the payload
 * source for the run-detail / debug panel.
 */
export const getWorkflowRun = (workflowId: string, runId: string): Promise<{
  success: boolean
  data?: WorkflowRun
  message?: string
}> => get(`/api/v1/workflows/${workflowId}/runs/${runId}`)

/** Path-only SSE URL; the stream composable adds base URL + auth headers. */
export function workflowRunEventsUrl(workflowId: string, runId: string): string {
  return `/api/v1/workflows/${workflowId}/runs/${runId}/events`
}

File: internal/types/workflow.go
package types

import (
	"time"

	"gorm.io/gorm"
)

// Workflow status lifecycle values. A workflow starts as draft, is
// published when its DSL is considered stable, and can be archived to
// hide it from editors without deleting run history.
const (
	WorkflowStatusDraft     = "draft"
	WorkflowStatusPublished = "published"
	WorkflowStatusArchived  = "archived"
)

// ValidWorkflowStatuses is the closed set of accepted workflow.status values.
var ValidWorkflowStatuses = []string{WorkflowStatusDraft, WorkflowStatusPublished, WorkflowStatusArchived}

// IsValidWorkflowStatus reports whether status is one of ValidWorkflowStatuses.
func IsValidWorkflowStatus(status string) bool {
	switch status {
	case WorkflowStatusDraft, WorkflowStatusPublished, WorkflowStatusArchived:
		return true
	default:
		return false
	}
}

// CreateWorkflowRequest is the REST payload for POST /api/v1/workflows.
// DSL is the verbatim workflow document; status defaults to draft.
type CreateWorkflowRequest struct {
	Name        string `json:"name" binding:"required"`
	Description string `json:"description"`
	DSL         JSON   `json:"dsl"`
	Status      string `json:"status"` // optional; empty = draft
}

// UpdateWorkflowRequest is the REST payload for PUT /api/v1/workflows/:id.
// The update is a full replace of the mutable fields: omitted fields reset
// to their zero value. Bump of `version` is server-side.
type UpdateWorkflowRequest struct {
	Name        string `json:"name" binding:"required"`
	Description string `json:"description"`
	DSL         JSON   `json:"dsl"`
	Status      string `json:"status"` // optional; empty = keep current
}

// Workflow represents a tenant-scoped workflow orchestration definition.
//
// The DSL document is stored verbatim (types.JSON). It carries a dual view —
// `graph` (nodes/edges for canvas rendering) and `components` (upstream/
// downstream execution topology) — whose structural validation lives in the
// service layer; this type intentionally treats it as opaque.
type Workflow struct {
	// Unique identifier of the workflow (UUID, generated in Go)
	ID string `yaml:"id" json:"id" gorm:"type:varchar(36);primaryKey"`
	// Tenant ID (data-isolation scope; every repository query filters on it)
	TenantID uint64 `yaml:"tenant_id" json:"tenant_id" gorm:"not null;index"`
	// Creator user ID; empty means tenant-owned (Admin+ only to mutate)
	CreatorID string `yaml:"creator_id" json:"creator_id" gorm:"type:varchar(36);not null;default:''"`
	// Display name (1-255 characters)
	Name string `yaml:"name" json:"name" gorm:"type:varchar(255);not null"`
	// Free-form description (<= 2000 characters)
	Description string `yaml:"description" json:"description" gorm:"type:text"`
	// Workflow DSL document, stored verbatim (dual view: graph + components)
	DSL JSON `yaml:"dsl" json:"dsl" gorm:"type:jsonb;not null;default:'{}'"`
	// Lifecycle status: draft | published | archived
	Status string `yaml:"status" json:"status" gorm:"type:varchar(50);not null;default:'draft'"`
	// bumped by 1 on every successful update
	Version int `yaml:"version" json:"version" gorm:"not null;default:1"`

	CreatedAt time.Time      `yaml:"created_at" json:"created_at"`
	UpdatedAt time.Time      `yaml:"updated_at" json:"updated_at"`
	DeletedAt gorm.DeletedAt `yaml:"deleted_at" json:"deleted_at"`
}

// WorkflowRun represents one execution record of a workflow.
//
// Populated by the execution-wiring slice; until then no endpoint creates
// rows and the list endpoint returns an empty history.
// RunWorkflowRequest is the input document of one workflow execution.
type RunWorkflowRequest struct {
	// Query is materialized into sys.query by the Start node.
	Query string `json:"query"`
	// Files (optional) are materialized into sys.files.
	Files []string `json:"files,omitempty"`
	// Async selects the execution mode: true enqueues a workflow:run task
	// and returns the run row immediately in status=pending (HTTP 202);
	// false/omitted executes synchronously (HTTP 200, 120s cap).
	Async bool `json:"async,omitempty"`
}

// WorkflowRunEvent is one progress frame of a workflow run, delivered to
// SSE subscribers (per-run broker) and the global event bus (observability).
type WorkflowRunEvent struct {
	WorkflowID string `json:"workflow_id"`
	RunID      string `json:"run_id"`
	// Kind discriminates frames: "node" (node lifecycle) | "run" (terminal).
	Kind string `json:"kind"`
	// NodeID is set for Kind=node frames.
	NodeID string `json:"node_id,omitempty"`
	// Phase: node frames carry started|finished|failed; run frames carry the
	// terminal run status (succeeded|failed|cancelled).
	Phase string `json:"phase"`
	// Err is the terminal error message for failed phases.
	Err string `json:"error,omitempty"`
	// DurationMS is the node execution duration for finished/failed frames.
	DurationMS int64 `json:"duration_ms,omitempty"`
	// Status is the terminal run status for Kind=run frames.
	Status string `json:"status,omitempty"`
}

// Workflow run statuses (persisted in workflow_runs.status).
const (
	WorkflowRunStatusPending   = "pending"
	WorkflowRunStatusRunning   = "running"
	WorkflowRunStatusSucceeded = "succeeded"
	WorkflowRunStatusFailed    = "failed"
	WorkflowRunStatusCancelled = "cancelled"
)

type WorkflowRun struct {
	// Unique identifier of the run (UUID, generated in Go)
	ID string `yaml:"id" json:"id" gorm:"type:varchar(36);primaryKey"`
	// Tenant ID (data-isolation scope)
	TenantID uint64 `yaml:"tenant_id" json:"tenant_id" gorm:"not null;index:idx_workflow_runs_tenant_workflow"`
	// Owning workflow ID (paired with TenantID in the composite index)
	WorkflowID string `yaml:"workflow_id" json:"workflow_id" gorm:"type:varchar(36);not null;index:idx_workflow_runs_tenant_workflow"`
	// Run status: pending | running | succeeded | failed | cancelled
	Status string `yaml:"status" json:"status" gorm:"type:varchar(50);not null;default:'pending'"`
	// Run input document (opaque JSON)
	Input JSON `yaml:"input" json:"input" gorm:"type:jsonb"`
	// Run output document (opaque JSON)
	Output JSON `yaml:"output" json:"output" gorm:"type:jsonb"`
	// Terminal error message when status=failed
	Error string `yaml:"error" json:"error" gorm:"type:text"`

	CreatedAt time.Time      `yaml:"created_at" json:"created_at"`
	UpdatedAt time.Time      `yaml:"updated_at" json:"updated_at"`
	DeletedAt gorm.DeletedAt `yaml:"deleted_at" json:"deleted_at"`
}
